router/broker.go

```diff
@@ -305,47 +362,80 @@ func (b *broker) syncPublish(pub *wamp.Session, msg *wamp.Publish, pubID wamp.ID
 }
 
 func newSubscription(id wamp.ID, subscriber *wamp.Session, topic wamp.URI, match string) *subscription {
+	subscribers := map[*wamp.Session]struct{}{}
+	if subscriber != nil {
+		subscribers = map[*wamp.Session]struct{}{subscriber: {}}
+	}
+
 	return &subscription{
 		id:          id,
 		topic:       topic,
 		match:       match,
 		created:     wamp.NowISO8601(),
-		subscribers: map[*wamp.Session]struct{}{subscriber: {}},
+		subscribers: subscribers,
 	}
 }
 
-func (b *broker) syncSubscribe(subscriber *wamp.Session, msg *wamp.Subscribe, match string) {
+func (b *broker) syncSaveEvent(eventStore *eventHistoryStoreItem, pub *wamp.Publish, event *wamp.Event) {
+	item := &eventHistoryItem{
+		publication: pub,
+		event: &storedEvent{
+			timestamp:    time.Now(),
+			Subscription: event.Subscription,
+			Publication:  event.Publication,
+			Details:      event.Details,
+			Arguments:    event.Arguments,
+			ArgumentsKw:  event.ArgumentsKw,
+		},
+	}
+	eventStore.events = append(eventStore.events, item)
+
+	if eventStore.isLimitReached {
+		eventStore.events = eventStore.events[1:]
```

Comment:
This is okay but will quickly recycle a lot of memory. What you could do is to make a ring buffer of events and without causing any GC churn, keep the last N events. A pre-type-parameter `ring.New` function will give you a ring buffer for N values.

Reply:
Though on second though, I do not recommend using `ring.Ring`, as it forces you to go through `any`, making storage inefficient.